### DISCUSSION TITLE
Direct I/O with Gramine

### DISCUSSION THREAD

**Discussion by ila** [gramineproject/gramine#1801] [Performance discussions]
Hi, we are benchmarking a database system on SGX, where our workload is primarily to read from disk-based (encrypted) files. 

We know that there is no caching of computations output, either by the application or the OS. Only the input files are cached, in case of buffered I/O. Therefore, as a way to reduce the overhead of page swapping (incurred running many different huge queries consecutively), we decided to try direct I/O. 

However, both `gramine-direct` and `gramine-sgx` do not seem to be affected by direct IO. According to the documentation, `O_DIRECT` is not supported.

 Does this mean that there is no way in Gramine to perform direct I/O? 

Is there a way to implement this syscall, or to forcefully avoid caching of I/O? 

Is directed I/O a safe operation to perform inside SGX, what is the reasoning behind ignoring this flag?

Thanks in advance!

**Comment by dimakuv** [comment]
You are correct. This flag is **ignored** by Gramine: https://gramine.readthedocs.io/en/stable/devel/features.html#file-system-operations.

(By the way, ignoring the `O_DIRECT` flag is considered a legit behavior. `O_DIRECT` is more like a hint, but its exact semantics depend on a number of factors. See here: https://stackoverflow.com/questions/41257656/what-does-o-direct-really-mean)

This means that the host Linux will **not receive this flag** during opening of the file.

> Is directed I/O a safe operation to perform inside SGX, what is the reasoning behind ignoring this flag?

I think there is no reasoning behind ignoring this flag. I don't remember anyone ever doing such an experiment and asking us this question. In other words: by default Gramine forbids/ignores most flags/features, just to be on the safe side. When a customer comes and asks for a particular flag/feature, then we consider implementing it. So the reason is purely "because nobody cared before".

Now that you raise this question, it seems to me that simply forwarding `O_DIRECT` from the app through Gramine and to the host OS is a security-unrelated choice. In the best (benign) case, the host OS will honor this flag and you'll see performance improvement. In the worst (malicious) case, the host OS will ignore this flag and you'll see bad performance. The latter case is a variant of Denial of Service, which is not in the threat model of Intel SGX and Gramine. So if I'm not mistaken, implementing `O_DIRECT` is simple and safe.

I will educate myself a bit on `O_DIRECT` and its consequences, and if all is well, I'll submit a PR to add this flag.

**Reply by ila** [reply]
Thanks a lot for this very insighful explaination :)

**Comment by dimakuv** [comment]
@ila Quick question: according to your experiments, what is the performance improvement when using `O_DIRECT` on your original (native) database application? Just curious to know how beneficial this feature is.

**Reply by ila** [reply]
In our native database application, running outside SGX, I am not sure it makes too much sense to evaluate this kind of behaviour - usually, database applications highly benefit from data being in cache, and buffered IO is the preferred solution. In general, I'd say the IO time is slightly slower (due to having to read blocks of fixed size, so allocating more memory and copying if the data does not align). However, in SGX, in huge workloads, a lot of page swapping occurs - the OS filesystem will cache pages that the enclave is not currently using, and the enclave will fetch them from there. This creates a lot of overhead since the pages need to be encrypted and decrypted and page faults cause CPU interrupts. We are confident IO_DIRECT is going to fix these issues - apparently the overhead of I/O is still less significant than the constant page swapping. Let me know if you have any more questions :)

just to add a bit more context - we are experiencing up to 50x slowdown of queries (60GB dataset) if we run them consecutively in an enclave, compared to running them separately in a newly-initialized enclave. We think all of this is due to page swapping - tuning parameters such as the enclave size and the available DB memory help, but still 2-3x slowdown compared to the "optimal" performance

**Comment by dimakuv** [comment]
@ila Can you explain a bit more about your scenario:

1. Are you using `fs.mounts = { type="encrypted" }` for your DB files that you want to apply `O_DIRECT` on?
2. Have you tried changing from Encrypted files to Allowed files, i.e. declaring DB files as `sgx.allowed_files`? Do you still observe high overhead in that case?

Update from my side: I hacked a Gramine version that applies `O_DIRECT` on allowed files. However, it is **not** currently possible to apply this to encrypted files, because encrypted files' read/write logic is more complicated and uses bounce buffers. And `O_DIRECT` has this specific requirement to allocate memory in file-block-aligned memory regions (which would require significant changes to Gramine codebase, and is also somehow brittle by itself).

Reading more on `O_DIRECT`, this flag/feature is really controversial... I'm now leaning towards **not** having it in Gramine, unless there are significant workloads and benchmarks that prove otherwise.

I continue reading about `O_DIRECT` and whether it can be applied at the level of the host OS (so we can toggle it on Linux and leave Gramine unmodified). I'll update my reply when I'm done.

**Reply by dimakuv** [reply]
I don't want to create a branch/draft PR for my current hack, so I'll post what I've done for Allowed files here:
```diff
commit 5d84effd0ec8f178de012461accbda144d2e74e6 (HEAD -> dimakuv/add-odirect-flag)
Author: Dmitrii Kuvaiskii <dmitrii.kuvaiskii@intel.com>
Date:   Fri Mar 8 11:08:50 2024 -0800

    WIP [LibOS,PAL] Add passthrough support for `O_DIRECT` flag during file open

    This flag may be beneficial in some corner cases, to avoid page swapping
    due to buffered I/O (page swapping is especially expensive for SGX
    enclaves). From the security perspective, `O_DIRECT` is a host-OS
    performance optimization and thus the worst-case scenario is malicious
    OS ignoring this flag, i.e. Denial of Service.

    WIP: Currently works only for Allowed Files. Not implemented for
    Encrypted Files, as `O_DIRECT` requires file-blocked-aligned memory
    regions, which are *not* possible in current Encrypted Files impl.

    Signed-off-by: Dmitrii Kuvaiskii <dmitrii.kuvaiskii@intel.com>

diff --git a/libos/include/libos_flags_conv.h b/libos/include/libos_flags_conv.h
index 865b8b49..603402e7 100644
--- a/libos/include/libos_flags_conv.h
+++ b/libos/include/libos_flags_conv.h
@@ -58,5 +58,8 @@ static inline enum pal_create_mode LINUX_OPEN_FLAGS_TO_PAL_CREATE(int flags) {
 }

 static inline pal_stream_options_t LINUX_OPEN_FLAGS_TO_PAL_OPTIONS(int flags) {
-    return flags & O_NONBLOCK ? PAL_OPTION_NONBLOCK : 0;
+    pal_stream_options_t pal_options = 0;
+    pal_options |= flags & O_NONBLOCK ? PAL_OPTION_NONBLOCK : 0;
+    pal_options |= flags & O_DIRECT ? PAL_OPTION_DIRECT : 0;
+    return pal_options;
 }
diff --git a/libos/src/libos_parser.c b/libos/src/libos_parser.c
index e9de40c9..b2c8d3b7 100644
--- a/libos/src/libos_parser.c
+++ b/libos/src/libos_parser.c
@@ -722,6 +722,10 @@ static void parse_open_flags(struct print_buf* buf, va_list* ap) {
         buf_puts(buf, "|O_EXCL");
         flags &= ~O_EXCL;
     }
+    if (flags & O_DIRECT) {
+        buf_puts(buf, "|O_DIRECT");
+        flags &= ~O_DIRECT;
+    }

     if (flags)
         buf_printf(buf, "|0x%x", flags);
diff --git a/pal/include/host/linux-common/pal_flags_conv.h b/pal/include/host/linux-common/pal_flags_conv.h
index 6cf6d0a3..1d04d327 100644
--- a/pal/include/host/linux-common/pal_flags_conv.h
+++ b/pal/include/host/linux-common/pal_flags_conv.h
@@ -60,6 +60,11 @@ static inline int PAL_CREATE_TO_LINUX_OPEN(enum pal_create_mode create) {
 }

 static inline int PAL_OPTION_TO_LINUX_OPEN(pal_stream_options_t options) {
-    assert(WITHIN_MASK(options, PAL_OPTION_NONBLOCK | PAL_OPTION_PASSTHROUGH));
-    return options & PAL_OPTION_NONBLOCK ? O_NONBLOCK : 0;
+    assert(WITHIN_MASK(options, PAL_OPTION_NONBLOCK |
+                                PAL_OPTION_PASSTHROUGH |
+                                PAL_OPTION_DIRECT));
+    int linux_options = 0;
+    linux_options |= options & PAL_OPTION_NONBLOCK ? O_NONBLOCK : 0;
+    linux_options |= options & PAL_OPTION_DIRECT ? O_DIRECT : 0;
+    return linux_options;
 }
diff --git a/pal/include/pal/pal.h b/pal/include/pal/pal.h
index 259082fb..7b2bbe81 100644
--- a/pal/include/pal/pal.h
+++ b/pal/include/pal/pal.h
@@ -326,7 +326,8 @@ typedef uint32_t pal_stream_options_t; /* bitfield */
 #define PAL_OPTION_EFD_SEMAPHORE   0x1 /*!< specific to `eventfd` syscall */
 #define PAL_OPTION_NONBLOCK        0x2
 #define PAL_OPTION_PASSTHROUGH     0x4 /*!< Disregard `sgx.{allowed,trusted}_files` */
-#define PAL_OPTION_MASK            0x7
+#define PAL_OPTION_DIRECT          0x8 /*!< corresponds to `O_DIRECT`, doesn't affect security */
+#define PAL_OPTION_MASK            0xf

 /*!
  * \brief Open/create a stream resource specified by `uri`.
```

I tested with a random example I found on the internet: https://gist.github.com/ddk50/987c645b0f476e9449eb9786a9bf8c4d

**Reply by ila** [reply]
@dimakuv again, many thanks for taking your time to help us out.
1) we are not using encrypted fs mounts - we are mounting directories in an unencrypted ways and placing the specific applicatios/libraries in the trusted files (database executables and dependencies).
2) hovwever, the database files (ie the ones containing the data itself) are in allowed files indeed - our reasoning behind this is that if these files are encrypted to begin with, it is "safe" for the application to access them, as it needs a secret key to open and modify them. Mind you, the encryption I am mentioning is entirely application based (in short, we take Parquet files and we encrypt block by block with AES). This with io_direct has been tested outside Gramine and seems to be working fine.

Therefore, the scenario you are describing is pretty much already the case we are using to benchmark our queries. Which is great, because I can immediately test your code :)

I have a couple more questions now:
1) how come the encrypted file logic is more difficult to handle? Please pardon my ignorance, but I thought that encryption/decryption happens at a higher level than read/write operations, so what do you think is the issue here? In our application, since we encrypt/decrypt data in memory before/after I/O, we think it won't make a difference. However, I am not familiar with the encryption/decryption system of Gramine, so I am curious to see whether it works differently :)
2) How come the feature is controversial? I totally understand if it would require huge changes to the codebase, I am more interested about the security/performance/stability issues here

Also thanks a lot for this patch! Please don't feel compelled to fix all our issues - my aim here is not to convince you, but rather to understand :) and if this doesn't work out, we will find another way to optimize performance in Gramine or modify our application accordingly. I appreciate a lot your input and availability answering these doubts.

**Reply by mkow** [reply]
> hovwever, the database files (ie the ones containing the data itself) are in allowed files indeed - our reasoning behind this is that if these files are encrypted to begin with, it is "safe" for the application to access them, as it needs a secret key to open and modify them. Mind you, the encryption I am mentioning is entirely application based (in short, we take Parquet files and we encrypt block by block with AES). This with io_direct has been tested outside Gramine and seems to be working fine.

Sorry for a slight off-topic, but a warning here: this is most likely insecure and it's possible to modify the database from the host without the app noticing. It depends on the encryption you're using, but most likely your scheme is similar to the disk encryption ones, which are only protecting against data leaking, but not against an active attacker who can modify the data. You'd need to either reencrypt the whole file on each modification or use a merkle tree for a more localized integrity checks. What encryption scheme are you using exactly?

**Reply by ila** [reply]
We are using either AES-GCM from the OpenSSL library or the same from mbedtls (currently testing different methods - if you have any suggestions, we are open to them - we are trying to exploit Intel-based hardware instructions in Gramine). Our key can be a variable size. We are aware that data modifications would need a re-encryption of the whole database file, therefore right now our workloads are read-only (which is acceptable for analytical workloads). The data is only unencrypted when it resides in the enclave memory. If edits would be possible, we would need to 1) decrypt the whole database inside the enclave, 2) apply modifications, and 3) re-encrypt it before writing on disk. We are still working on this part, so if there are security issues here, we are happy to address them in the future.

**Reply by dimakuv** [reply]
> how come the encrypted file logic is more difficult to handle? Please pardon my ignorance, but I thought that encryption/decryption happens at a higher level than read/write operations, so what do you think is the issue here?

Since you're using `sgx.allowed_files` for your database, the discussion becomes redundant. But just to satisfy your curiosity:
1. At least on my current Linux with ext4 filesystem, a `write()` from a non-aligned memory address returns with `-EINVAL` (invalid value error). This means that the application must allocate buffers in a special way (e.g. using `posix_memalign()`).
2. Recall that Gramine's Encrypted Files are transparently encrypted and decrypted inside the SGX enclave. This necessitates a "bounce buffer" -- a temporary buffer that sits in-between the user buffer (the buffer in your program) and the host-OS buffer. Given requirement above, this temporary buffer must be allocated with a correct alignment.
3. Gramine has its own memory allocator (slab allocator) that doesn't have an alignment specifier. So we would need to modify our memory allocator to satisfy the requirement above. This is extremely painful.
4. I am unclear what are all the alignment requirements -- it looks like every OS and every FS implement their own alignment requirements. So how should Gramine choose the default alignment? How can Gramine be tuned to set an alignment for a particular program? It just gets annoying with all this "`O_DIRECT` is not standardized".

> How come the feature is controversial? I totally understand if it would require huge changes to the codebase, I am more interested about the security/performance/stability issues here

Well, `O_DIRECT` was never standardized from what I know. Some FSes support it, some silently ignore it, some try to support but silently fall back to buffered I/O, some have specific requirements and fail if the requirements are not fullfilled, etc. Check these (old but seem to hold true):
- https://ext4.wiki.kernel.org/index.php/Clarifying_Direct_IO%27s_Semantics
- https://yarchive.net/comp/linux/o_direct.html

**Reply by mkow** [reply]
> We are using either AES-GCM from the OpenSSL library or the same from mbedtls (currently testing different methods - if you have any suggestions, we are open to them - we are trying to exploit Intel-based hardware instructions in Gramine).

AES is of course good, it's mostly about the higher-level scheme you're using (and these mostly aren't directly implemented in hardware). In this case it's GCM, which is good as long as you don't want partial modifications to the file.

> Our key can be a variable size. We are aware that data modifications would need a re-encryption of the whole database file, therefore right now our workloads are read-only (which is acceptable for analytical workloads). The data is only unencrypted when it resides in the enclave memory.

Ok, then it sounds good :)

> If edits would be possible, we would need to 1) decrypt the whole database inside the enclave, 2) apply modifications, and 3) re-encrypt it before writing on disk. We are still working on this part, so if there are security issues here, we are happy to address them in the future.

This design is good (except performance, as you need to re-encrypt the whole database), just remember that it will still be possible to rollback the whole database to an older version (the untrusted host can save an older encrypted database and overwrite a newer state with it). But this is fine in most cases, as long as the rollback can only be done to the whole database, not e.g. only to a specific encryption block.

**Reply by ila** [reply]
Thank you so much for this explaination! It is very interesting for us to know about the potential performance and security issues (trying to write a paper about analytical workloads on SGX), so all this information is very precious to extend our knowledge and implementative choices.

**Comment by ila** [comment]
Edit: your patch compiles and builds no problem, however I am not seeing any performance difference than we had before. The steps I did are:
- uninstalled Gramine, cloned from source
- applied your changes (4 files)
- built from scratch using the instructions in the documentation
- cleaned our Gramine-generated manifest and built it again
- tested with `gramine-direct`

It could also be that any of these steps somehow did not work, but it is hard to tell since I did test in between steps that the `gramine` executable did not exist. I am using the upstream SGX drivers (already preinstalled in our Intel machine, Intel Xeon Platinum 8352Y).

To give you an example, this is the query we are running:
`Running query: SELECT UserID FROM read_parquet('duckdb-parquet-encrypted/hits_encrypted.parquet', encryption_config={footer_key: 'key128'}, force_direct_io=true);`
1st run: 2.82s
2nd run: 1.78s
3rd run: 1.79s

The fact that there is a difference between the 1st and the successive runs makes me guess that the IO_DIRECT flag was still ignored. Running the very same query outside Gramine gives us 3 equal results, so we don't think it's an issue on your side. 

Just to clarify, the expected behaviour is **3 equal results for 3 runs**, which imply that nothing is being cached, therefore our IO is not buffered. To give you a reference, this is how we are implementing IO_DIRECT in our application:
https://github.com/duckdb/duckdb/pull/11028/commits/a358dfad4b5178e4b6c56416bf5e345d0dd7782d
https://github.com/duckdb/duckdb/pull/11028/commits/bf93e1c49c65192a5a40fbcec68dbb1f919eb571

Edit: I also tried the very same workload without file encryption, and the behaviour is still the same (1st run slower than 2nd and 3rd). So, we can rule out application-based encryption as the culprit).

I have taken a look at the file you sent, how do you make sure the behaviour is indeed correct? Should it give a different output on `gramine-direct` compiled with and without the IO_DIRECT hack? We are purely comparing performance here, so it could also be that the IO_DIRECT works and something else is being cached for some reason (but it seems unlikely, as this behaviour is very reproducible). If you need some example to reproduce, I can provide that as well. Have a good day :)

Another edit: this is our manifest file if needed: https://github.com/cwida/ClickBench-SGX/blob/main/benchmark.manifest.template
You can check the README and the correspondent folders to have an idea of what we are doing, although for full reproducibility a few extra steps are needed (wouldn't recommend, these are going to take some time, but can help if needed).

**Reply by mkow** [reply]
Could you `strace` all the gramine processes (i.e., in follow child mode) and check if `open` on the host is actually called with `O_DIRECT`?

**Reply by dimakuv** [reply]
Yep, exactly what @mkow said.

That's how I tested that my patch works (on a trivial program that I also linked in my prevous message).

**Reply by ila** [reply]
Thanks for the pointer! This is the command I use:

`strace gramine-direct benchmark duckdb/profile_duckdb.py 2>&1 | grep open`
Output:
```
open("./duckdb-parquet-encrypted/hits.parquet", O_RDONLY|O_CLOEXEC) = 278
open("./duckdb-parquet-encrypted/hits.parquet", O_RDONLY|O_CLOEXEC) = 278
open("./duckdb-parquet-encrypted/hits.parquet", O_RDONLY|O_CLOEXEC) = 278
open("./duckdb-parquet-encrypted/hits.parquet", O_RDONLY|O_DIRECT|O_CLOEXEC) = 286
open("./duckdb-parquet-encrypted/hits.parquet", O_RDONLY|O_CLOEXEC) = 278
open("./duckdb-parquet-encrypted/hits.parquet", O_RDONLY|O_CLOEXEC) = 278
open("./duckdb-parquet-encrypted/hits.parquet", O_RDONLY|O_CLOEXEC) = 278
open("./duckdb-parquet-encrypted/hits.parquet", O_RDONLY|O_DIRECT|O_CLOEXEC) = 291
open("./duckdb-parquet-encrypted/hits.parquet", O_RDONLY|O_CLOEXEC) = 278
open("./duckdb-parquet-encrypted/hits.parquet", O_RDONLY|O_CLOEXEC) = 278
open("./duckdb-parquet-encrypted/hits.parquet", O_RDONLY|O_CLOEXEC) = 278
open("./duckdb-parquet-encrypted/hits.parquet", O_RDONLY|O_DIRECT|O_CLOEXEC) = 309
open("./duckdb/my-db.duckdb.wal", O_RDONLY|O_CLOEXEC) = 149
```
I see a similar `strace` output running my application outside Gramine. On the other hand, if I try the very same command on the official Gramine installation, the O_DIRECT flag is missing. So I am guessing the O_DIRECT flag is working? Could the performance slowdown be attributed to something else then?

**Reply by mkow** [reply]
Could you also print read/readv/etc. syscalls? I wonder if they are called on these `O_DIRECT` descriptors or non-`O_DIRECT` ones.

**Reply by ila** [reply]
I uploaded the `strace` output (custom Gramine build with this patch applied) to a file: https://github.com/cwida/ClickBench-SGX/blob/main/strace_log.txt

This is the "baseline" log (running the very same thing outside Gramine) if that helps: https://github.com/cwida/ClickBench-SGX/blob/main/strace_log_baseline.txt

Also, it seems like some of these calls (`open`) on the `hits.parquet` file are with O_DIRECT, some are not. Furthermore, some of them are missing the O_SYNC flag - maybe that is the problem?

**Reply by dimakuv** [reply]
> This is the "baseline" log (running the very same thing outside Gramine) if that helps: https://github.com/cwida/ClickBench-SGX/blob/main/strace_log_baseline.txt

The logs seem to use the FDs (file descriptors) for `pread()` as they are opened with via `open(O_DIRECT)`. Looks good to me, which seems to disprove your theory about the effect of `O_DIRECT`...

> Also, it seems like some of these calls (`open`) on the `hits.parquet` file are with O_DIRECT, some are not. Is this an expected behaviour?

That's expected behavior. Gramine sometimes needs to do "helper" opens of the file, to learn some additional meta-information on it.

> Furthermore, some of them are missing the O_SYNC flag - maybe that is the problem?

Ok, **now this** may be a real problem. I remember reading that `O_DIRECT` may be useless (= ignored) if `O_SYNC` flag is not additionally specified.

**Reply by mkow** [reply]
> Also, it seems like some of these calls (`open`) on the `hits.parquet` file are with O_DIRECT, some are not. Is this an expected behaviour?

That's exactly why I asked about logs with `read()`, let's see... (maybe only the FDs opened with `O_DIRECT` are used for large reads)

> ```
> pread64(278, "\2248%\0PAR1", 8, 14779976438) = 8
> ```

Again slightly off-topic, but seems you don't encrypt metadata? Why?

> ```
> open("./duckdb-parquet-encrypted/hits.parquet", O_RDONLY|O_CLOEXEC) = 278
> [...]
> pread64(278, "\2248%\0PAR1", 8, 14779976438) = 8
> pread64(278, "\25\2\31\374j5\0\30\6schema\25\322\1\0\25\4%\0\30\7WatchID"..., 2439316, 14777537122) = 2439316
> ```

There's at least one large read without O_DIRECT here...
So, it's either @dimakuv's patch which is incomplete, or your app is not always using `O_DIRECT` when opening the database.

**Reply by mkow** [reply]
Ok, I see in the other logs that the app is not using `O_DIRECT` for this particular read, but later it starts to. Dmitrii's observation could also be the culprit, the patch doesn't forward `O_SYNC` to the host.

**Reply by ila** [reply]
- About O_SYNC: could this be an issue? Once again I am reading a lot of controversial information about this flag (now I understand why you don't want it there...), as far as I can tell it should be an issue only writing, or on MacBooks, but I could be wrong --- would implementing this fix this issue? I could even try myself if the extent of the patch is the same as O_DIRECT
- About large reads and metadata: the whole Parquet file should be opened with O_DIRECT, but indeed metadata does not use O_DIRECT (this is intended because at the moment we only wanted to test the file read performance, we plan to add this in the future). Metadata should be in two (small) files, so that could explain the weird syscalls. However performance results indicate a negligible performance difference with buffered metadata, so we don't think it might be the issue here

**Reply by ila** [reply]
Edit: our application implementation does indeed use O_SYNC in combination with O_DIRECT (on Linux), see https://github.com/duckdb/duckdb/blob/5a59c56acf2cf8eb1948e9045796019f781b317b/src/common/local_file_system.cpp#L313

**Reply by dimakuv** [reply]
> would implementing this fix this issue? I could even try myself if the extent of the patch is the same as O_DIRECT

Yes, implementing ~~`O_ASYNC`~~(`O_SYNC`) will basically be the same as `O_DIRECT`. Feel free to code this.

**Reply by ila** [reply]
On it - just a clarification, you mean `O_SYNC` here right? `O_ASYNC` is something different as far as I understand.

**Reply by ila** [reply]
Update - implemented and tested it, now the `strace` logs for my application inside and outside Gramine are the same, for example:
```
strace gramine-direct benchmark duckdb/profile_duckdb.py 2>&1 | grep O_SYNC
open("./duckdb-parquet-encrypted/hits.parquet", O_RDONLY|O_SYNC|O_DIRECT|O_CLOEXEC) = 297
open("./duckdb-parquet-encrypted/hits.parquet", O_RDONLY|O_SYNC|O_DIRECT|O_CLOEXEC) = 317
open("./duckdb-parquet-encrypted/hits.parquet", O_RDONLY|O_SYNC|O_DIRECT|O_CLOEXEC) = 280
```
However, performance is still unaffected. So the issue here could be:
1) either the metadata files are buffered and that causes a different performance, although I doubt that only that causes such a big improvement (1s performance difference for 14GB is quite a bit)
2) Gramine (or the OS) is caching something that we are unaware of
3) the implementation of these syscalls is problematic (or maybe our implementation is problematic)

I'll look more into it, maybe my advisor has some ideas. If you have any other suggestions, I am open :) in the meantime, again, many thanks for the help!

**Reply by dimakuv** [reply]
@ila Have you tried `gramine-direct`? If you see significant overhead in `gramine-direct`, then please debug/profile this first.

**Reply by ila** [reply]
Yes, all these logs are using `gramine-direct`.

Update: I actually tested with and without direct io, and it looks like the direct io is actually being applied - however there is still a bit of "noise" (200-300ms of difference between cold and hot runs), so I was wondering, is Gramine caching anything (regardless of the application) during its execution? 

Also, I now tested your code with `gramine-sgx` (only the patch you posted here, without my modifications) and it doesn't seem to work. The error it gives me is:
`Error IO Error: Could not read from file "duckdb-parquet-encrypted/hits.parquet": Invalid argument`

The very same thing with `gramine-direct` works no problem. Did you by any chance test with SGX? I feel like there might be something missing in the code.

**Reply by dimakuv** [reply]
@ila I did test with `gramine-sgx` too, and it worked. At least on my dummy example.

But this feels like a problem of alignments (`O_DIRECT` on ext4 requires 512B alignment, iiuc). I got the same error when I used `malloc()` (without alignment guarantees) vs `posix_memalign()` (with specific alignment guarantees).

I can imagine that my patch doesn't account for some corner case, with e.g. much bigger workloads and memory regions.

Anyway, it feels like the problem is **not** in `O_DIRECT` support, but somewhere else. Could you continue your performance analysis? For example, you can do `perf record` on `gramine-direct` (built in `debug_optimized` mode). You can find additional documentation on this here: https://gramine.readthedocs.io/en/stable/performance.html#profiling-with-perf. Of course, start with `gramine-direct` since already there you observe problems.

**Reply by ila** [reply]
Hi, just a quick update - we decided to avoid hacky stuff and try moving the application part dealing with I/O outside the enclave. This should help paging as well. Will keep you posted :)
